clang/lib/CIR/CodeGen/CIRGenCall.cpp

```diff
@@ -629,63 +465,32 @@ RValue CIRGenFunction::emitCall(const CIRGenFunctionInfo &CallInfo,
-         !RetAI.isCoerceAndExpand() && "NYI");
-
-  // When passing arguments using temporary allocas, we need to add the
-  // appropriate lifetime markers. This vector keeps track of all the lifetime
-  // markers that need to be ended right after the call.
-  assert(!cir::MissingFeatures::shouldEmitLifetimeMarkers() && "NYI");
-
-  // Translate all of the arguments as necessary to match the CIR lowering.
-  assert(CallInfo.arg_size() == CallArgs.size() &&
-         "Mismatch between function signature & arguments.");
   unsigned ArgNo = 0;
   CIRGenFunctionInfo::const_arg_iterator info_it = CallInfo.arg_begin();
   for (CallArgList::const_iterator I = CallArgs.begin(), E = CallArgs.end();
        I != E; ++I, ++info_it, ++ArgNo) {
-    const cir::ABIArgInfo &ArgInfo = info_it->info;
-
-    // Insert a padding argument to ensure proper alignment.
-    assert(!CIRFunctionArgs.hasPaddingArg(ArgNo) && "Padding args NYI");
-
-    unsigned FirstCIRArg, NumCIRArgs;
-    std::tie(FirstCIRArg, NumCIRArgs) = CIRFunctionArgs.getCIRArgs(ArgNo);
-
-    switch (ArgInfo.getKind()) {
-    case cir::ABIArgInfo::Direct: {
-      if (!mlir::isa<cir::RecordType>(ArgInfo.getCoerceToType()) &&
-          ArgInfo.getCoerceToType() == convertType(info_it->type) &&
-          ArgInfo.getDirectOffset() == 0) {
-        assert(NumCIRArgs == 1);
-        mlir::Value V;
-        assert(!I->isAggregate() && "Aggregate NYI");
-        V = I->getKnownRValue().getScalarVal();
-
-        assert(CallInfo.getExtParameterInfo(ArgNo).getABI() !=
-                   ParameterABI::SwiftErrorResult &&
-               "swift NYI");
-
-        // We might have to widen integers, but we should never truncate.
-        if (ArgInfo.getCoerceToType() != V.getType() &&
-            mlir::isa<cir::IntType>(V.getType()))
-          llvm_unreachable("NYI");
-
-        // If the argument doesn't match, perform a bitcast to coerce it. This
-        // can happen due to trivial type mismatches.
-        if (FirstCIRArg < CIRFuncTy.getNumInputs() &&
-            V.getType() != CIRFuncTy.getInput(FirstCIRArg))
-          V = builder.createBitcast(V, CIRFuncTy.getInput(FirstCIRArg));
-
-        CIRCallArgs[FirstCIRArg] = V;
-        break;
-      }
 
+    mlir::Type argType = convertType(info_it->type);
+    if (!mlir::isa<cir::RecordType>(argType)) {
+      mlir::Value V;
+      assert(!I->isAggregate() && "Aggregate NYI");
+      V = I->getKnownRValue().getScalarVal();
+
+      // We might have to widen integers, but we should never truncate.
+      if (argType != V.getType() && mlir::isa<cir::IntType>(V.getType()))
+        llvm_unreachable("NYI");
+
+      // If the argument doesn't match, perform a bitcast to coerce it. This
+      // can happen due to trivial type mismatches.
+      if (ArgNo < CIRFuncTy.getNumInputs() &&
+          V.getType() != CIRFuncTy.getInput(ArgNo))
+        V = builder.createBitcast(V, CIRFuncTy.getInput(ArgNo));
```

Comment:
I'm curious what would be the case when the argument's type does not match the function parameter's type. Won't Sema handle all the type mismatch cases?